storage/layers.go

```diff
@@ -2398,37 +2441,106 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
-	if !r.lockfile.IsReadWrite() {
-		return -1, fmt.Errorf("not allowed to modify layer contents at %q: %w", r.layerdir, ErrStoreIsReadOnly)
+func createTarSplitFile(r *layerStore, layerID string) (*os.File, error) {
+	if err := os.MkdirAll(filepath.Dir(r.tspath(layerID)), 0o700); err != nil {
+		return nil, err
 	}
+	return os.OpenFile(r.tspath(layerID), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o600)
+}
 
-	layer, ok := r.lookup(to)
-	if !ok {
-		return -1, ErrLayerUnknown
+// newMaybeStagedLayerExtraction initlaizes a new maybeStagedLayerExtraction. The caller
+// must call .cleanup() to remove any temporary files.
+func newMaybeStagedLayerExtraction(diff io.Reader, driver drivers.Driver) *maybeStagedLayerExtraction {
+	m := &maybeStagedLayerExtraction{
+		diff: diff,
 	}
+	if d, ok := driver.(drivers.ApplyDiffStaging); ok {
+		m.staging = d
+	}
+	return m
+}
 
+func (sl *maybeStagedLayerExtraction) cleanup() error {
+	return tempdir.CleanupTemporaryDirectories(sl.cleanupFuncs...)
+}
+
+// stageWithUnlockedStore stages the layer content without needing the store locked.
+// If the driver does not support stage addition then this is a NOP and does nothing.
+func (sl *maybeStagedLayerExtraction) stageWithUnlockedStore(r *layerStore, layerOptions *LayerOptions) error {
+	if sl.staging == nil {
+		// driver does not implement stage addition
+		return nil
+	}
+	td, err := tempdir.NewTempDir(filepath.Join(r.layerdir, tempDirPath))
+	if err != nil {
+		return err
+	}
+	sl.cleanupFuncs = append(sl.cleanupFuncs, td.Cleanup)
+
+	stageTarSplit, err := td.StageAddition()
+	if err != nil {
+		return err
+	}
+	sl.stagedTarSplit = stageTarSplit
+
+	f, err := os.OpenFile(stageTarSplit.Path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	result, err := applyDiff(layerOptions, sl.diff, f, func(payload io.Reader) (int64, error) {
+		cleanup, stageLayer, size, err := sl.staging.StartStagingDiffToApply(drivers.ApplyDiffOpts{
+			Diff:     payload,
+			Mappings: idtools.NewIDMappingsFromMaps(layerOptions.UIDMap, layerOptions.GIDMap),
+			// FIXME: What to do here? We have no lock and assigned label yet.
+			// Overlayfs should not need it anyway so this seems fine for now.
+			MountLabel: "",
+		})
+		sl.cleanupFuncs = append(sl.cleanupFuncs, cleanup)
+		sl.stagedLayer = stageLayer
+		return size, err
+	})
+	if err != nil {
+		return err
+	}
+
+	sl.result = result
+	return nil
+}
+
+// commitLayer() commits the content that was staged in stageWithUnlockedStore()
+//
+// Requires startWriting.
+func (sl *maybeStagedLayerExtraction) commitLayer(r *layerStore, layerID string) error {
+	err := sl.stagedTarSplit.Commit(r.tspath(layerID))
+	if err != nil {
+		return err
+	}
+	return sl.staging.CommitStagedLayer(layerID, sl.stagedLayer)
+}
+
+func applyDiff(layerOptions *LayerOptions, diff io.Reader, tarSplitFile *os.File, applyDriverFunc func(io.Reader) (int64, error)) (*applyDiffResult, error) {
```

Comment:
(Explicitly warning that this might be called without any layer store locks held would be nice.)